### PR TITLE
Dockerise CI testing and ensure existing test suite is passing

### PR DESCRIPTION
**Pull request by ajacombs**
_**Note:** Created as a pull request against `name_and_use_squash` branch to facilitate reviewing. Would suggest the large number of commits on this `ci/docker` branch are squashed / tidied prior to merging, and also note further work is needed before the changes are merged to the master branch._

This branch refactors the way tests are run locally and in GitHub Actions, so both pgTAP database tests and QGIS Python unit tests are run in docker containers. Additionally:
- QGIS unit tests have been updated so they all pass in QGIS 3.10, 3.16, and 3.24.
- pgTAP tests which were duplicated by Sqitch verification step have been removed, and all remaining pgTAP tests checked to be passing.
- The structure of the GitHub Actions config has been refactored so each step runs a small shell script, which can also be run locally from shell scripts which wrap these individual step scripts.
- Duplicated installation / automation scripts have been removed, with the intention that all local automation tasks are run from a series of small shell scripts in the `scripts` directory which can optionally call steps in the `scripts/steps` directory.
- Dockerfiles defining the PostgreSQL database containers used for testing are included, as well as a GitHub Actions config to automate their building and deployment to the GitHub Container Registry for this repo.

Changes of note which have not yet been made:
- I have not yet deleted either the `makefile` in the project root and the script at `db/scripts/nz-buildings-load`. These should both be made redundant, but I have not yet fully replaced their functionality as scripts in the `scripts` directory. Specifically these relate to local plugin installation, CHANGELOG updates, and new version deployments.
- The GitHub Actions config does not include all aspects of the previous version, specifically not including the parts for releasing versions. I've left this for now as I don't fully understand it and thought it warranted further discussion alongside the above point as to what the best way to update that functionality in new scripts is.
- I've made some changes to the main project README, but left in the parts relating to the above aspects which have not yet been changed, so it's not fully up to date on how to run tests etc (referencing things that have been partially but not fully removed yet).

### CODE REVIEW

File: buildings/tests/gui/test_processes_update_reference_data_admin_bdys.py [hunk 1]
```diff
@@ -42,98 +42,90 @@ def tearDown(self):
...
-        count_added = result.fetchone()[0]
-        self.assertEqual(count_added, 0)
-        # check suburb locality island with null suburb_4th is added
-        sql_added = (
-            "SELECT count(*)::integer FROM buildings_reference.suburb_locality WHERE external_suburb_locality_id = 108;"
-        )
-        result = db._execute(sql_added)
-        count_added = result.fetchone()[0]
-        self.assertEqual(count_added, 1)
+    # def test_suburb_locality_table_update(self):
```
Comment (pkinglinz) [inline buildings/tests/gui/test_processes_update_reference_data_admin_bdys.py:45]: are these hashed out sections needed? can we remove them completely?

File: buildings/tests/run.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+from pathlib import Path
+
+from qgis.core import QgsApplication
+from qgis.PyQt.QtCore import QDir
+from qgis.utils import iface
+
+from console.console_output import writeOut
+
+
+if iface is None:
```
Comment (pkinglinz) [inline buildings/tests/run.py:19]: perhaps include this in a try-except of the import of iface so is caught cleanly and returned with the error

File: buildings/utilities/database.py [hunk 3]
```diff
@@ -38,6 +38,7 @@
     _conn = psycopg2.connect(host=_host, port=_port, database=_dbname, user=_user, password=_pw)
 except psycopg2.DatabaseError as error:
     _conn = None
+    print(str(error))
```
Comment (pkinglinz) [inline buildings/utilities/database.py:41]: was this print statement just for you development? can it be removed

File: scripts/README.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
+# NZ Buildings scripts
```
Comment (pkinglinz) [inline scripts/README.md:1]: very nice well referenced readme, thankyou

File: scripts/steps/launch_database_container.sh [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+-e "PGHOST=$PGHOST" \
+-e "PGPORT=$PGPORT" \
+-e "PGUSER=$PGUSER" \
+-e "PGPASSWORD=$PGPASSWORD" \
+-e "POSTGRES_USER=$PGUSER" \
+-e "POSTGRES_PASSWORD=$PGPASSWORD" \
+-d \
+$DB_DOCKER_IMAGE
+
+RETRIES=3
```
Comment (pkinglinz) [inline scripts/steps/launch_database_container.sh:23]: would it be worthwhile adding a comment here as to why you are retrying as i thought this is worth knowing and could be a gotcha for folks in the future

**Review by pkinglinz — APPROVED** [review]
Looks good, really nice clean modules. I have left some comments but happy to approve as is. My .sh knowledge isn't great so good to have others looking over it

File: scripts/steps/sqitch_deploy_verify.sh [hunk 6]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
+
+echo >&2 ""
+echo >&2 "############################"
+echo >&2 "# Sqitch deploy and verify #"
+echo >&2 "############################deploy"
```
Comment (Douglas-K) [inline scripts/steps/sqitch_deploy_verify.sh:6]: Stray text 'deploy' that can be deleted

**Review by Douglas-K — APPROVED** [review]
Well done Andrew on making so many changes to modernise and simplify this as well as to get it working well.
I am unfamiliar with much of this code, but what I can understand looks good and your group session intro was helpful.  It will be good to have another session to provide updates since then.  I haven't tested this myself, but I am glad all automated tests pass!
Note that the Ubuntu LTR version was changed since you started working on this when it was 18.04, so this could be updated at some stage (but is minor as you say).
I only made one super duper minor inline comment as I was really clutching at straws to find something!

**Review by jducnuigeen — APPROVED** [review]
Full testing of changes on docker db with full restored database worked on a real bulk load and comparisons and publish.

**Comment by dwsilk** [comment]
Merged as part of https://github.com/linz/nz-buildings/pull/541.

**Pull request closed without merging**